Penumbra/UI/SharedTagManager.cs

```diff
@@ -0,0 +1,160 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Dalamud.Interface;
+using ImGuiNET;
+using OtterGui;
+using OtterGui.Raii;
+using OtterGui.Widgets;
+using Penumbra.Mods;
+using Penumbra.UI.Classes;
+
+namespace Penumbra.UI;
+public sealed class SharedTagManager
+{
+    private static uint _tagButtonAddColor = ColorId.SharedTagAdd.Value();
+    private static uint _tagButtonRemoveColor = ColorId.SharedTagRemove.Value();
+
+    private static float _minTagButtonWidth = 15;
+
+    private const string PopupContext = "SharedTagsPopup";
+    private bool _isPopupOpen = false;
+
+
+    public IReadOnlyList<string> SharedTags { get; internal set; } = Array.Empty<string>();
+
+    public SharedTagManager()
+    {
+    }
+
+    public void ChangeSharedTag(int tagIdx, string tag)
+    {
+        if (tagIdx < 0 || tagIdx > SharedTags.Count)
+            return;
+
+        if (tagIdx == SharedTags.Count) // Adding a new tag
+        {
+            SharedTags = SharedTags.Append(tag).Distinct().Where(tag => tag.Length > 0).OrderBy(a => a).ToArray();
```

Comment:
Having a list would make this here easier too, using Contains and Sort (or BinarySearch for efficiency if you keep it sorted anyway, I guess). There's also SortedList, but that's a dictionary so it's not quite what we need.
It might be worthwhile to make a small wrapper around the binarysearch and create a custom SortedList that's just a list with being sorted as an invariant.

Reply:
I stopped just a step shy of making a custom SortedList, and just made sure that the list gets sorted when loaded from .json, and that calls to `ChangeSharedTag()` keep the list sorted.

There's always the risk of somebody making code changes that edit the list in the future such that it's no longer sorted, but I think that's an acceptable risk given that it's not the end of the world if the list becomes un-sorted during runtime (and there's a few comments sprinkled about that should tip off the fact that it is supposed to be kept sorted).